[FLINK-6228][table] Integrating the OVER windows in the Table API (st…

sunjincheng121 opened the pull request:
In this PR I had integrating the OVER windows in the Table API, Implementation of the syntax and use examples are as follows：
* Syntax:
```
table
   .overWindows(
    (Rows|Range [ partitionBy value_expression , ... [ n ]] [ orderBy order_by_expression] 
      (preceding  UNBOUNDED|value_specification.(rows|milli|second|minute|hour|day|month|year)|CURRENTROW)
     [following UNBOUNDED|value_specification.(rows|milli|second|minute|hour|day|month|year)|CURRENTROW]
    as alias,...[n])
   )
  .select( [col1,...[n]], (agg(col1) OVER overWindowAlias, … [n])

```
* examples:
```
// Rows clause
table
   .window(Over partitionBy 'c orderBy 'rowTime  preceding 2.rows as 'w1)
   .select(
     'c,
     'b.count over 'w1 as 'countB,
     'e.sum over 'w1 as 'sumE)

// Range clause
table
   .window(Over partitionBy 'c orderBy 'rowTime preceding 2.milli as 'w1)
   .select(
     'c,
     'b.count over 'w1 as 'countB,
     'e.sum over 'w1 as 'sumE)
```
* More detail Info : https://docs.google.com/document/d/13Z-Ovx3jwtmzkSweJyGkMy751BouhuJ29Y1CTNZt2DI/edit#

NOTE: The documentation of the OVER tableAPI not included in this PR.
- [x] General
  - The pull request references the related JIRA issue ("[FLINK-6228][table] Integrating the OVER windows in the Table API")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/OverAggregate.scala:
```diff
@@ -23,7 +23,9 @@ import org.apache.calcite.rel.`type`.{RelDataType, RelDataTypeFieldImpl}
 import org.apache.calcite.rel.core.Window.Group
 import org.apache.calcite.rel.core.Window
-import org.apache.calcite.rex.{RexInputRef}
+import org.apache.calcite.rex.{RexInputRef, RexLiteral}
```
All these changes could be undone if we don't use a literal for the orderBy() expression

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/OverWindowITCase.scala:
```diff
@@ -0,0 +1,265 @@
+  }
+
+  @Test
+  def testRowTimeBoundedPartitionedRowOver(): Unit = {
```
Add a test for rowtime bounded range as well?

fhueske replied:
Fair enough. The `OverWindowTest` should be sufficient.
However, I would keep a few integration tests for the Table API as well.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/windows.scala:
```diff
@@ -22,6 +22,126 @@ import org.apache.flink.table.expressions.{Expression, ExpressionParser}
+    * @param following subsequent offset that relative to the current row
+    * @return this over window
+    */
+  def following(following: String): OverWindow = {
```
Should we make `following` optional and default to CURRENT_ROW / CURRENT_RANGE depending on the type of `preceding`? I think that would be a nice shortcut and also be aligned with SQL.

What do you think @sunjincheng121 ?

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/logical/operators.scala:
```diff
@@ -54,6 +54,9 @@ case class Project(projectList: Seq[NamedExpression], child: LogicalNode) extend
             case expr if !expr.valid => u
             case c @ Cast(ne: NamedExpression, tp) => Alias(c, s"${ne.name}-$tp")
             case gcf: GetCompositeField => Alias(gcf, gcf.aliasName().getOrElse(s"_c$i"))
+            case over: OverCall if null != over.aggAlias =>
```
If i understand you correctly, you want add alias for AGG,e.g.:
` .select('c, 'b.count over 'w as 'mycount) -->  .select('c, 'b.count  as 'mycount over 'w)`
 but for SQL we add alias to OVER, e.g.:
`SELECT c, count(a) OVER (PARTITION BY c ORDER BY ProcTime()  RANGE UNBOUNDED preceding) as cnt1 from T1`
So  I want to be consistent with SQL，What do you think?

fhueske replied:
Currently `.select('c, 'b.count over 'w as 'mycount)` would be translated into `OverCall(Count(b), w, mycount, _)`. If we remove `as` from `OverCall`, it will be wrapped in a regular `Alias` expression: `Alias(OverCall(Count(b), w, _), mycount)`.

The syntax would be the same, but the internal representation would use the existing `Alias` expression, just like any other expression that is renamed.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/call.scala:
```diff
@@ -49,6 +57,128 @@ case class Call(functionName: String, args: Seq[Expression]) extends Expression
+    overWindowAlias: Expression,
+    var overWindow: OverWindow = null) extends Expression {
+
+  private[flink] def as(aggAlias: Expression): OverCall = {
```
Please, tell me more about this. :)

fhueske replied:
If we remove the `as` method from `OverCall`, it will be handled as a regular`Alias` expression which wraps the `OverCall`. 
Since the `OverCall` is wrapped, the `ProjectTranslator.translateOverWindow()` method needs another `case Alias(OverCall(agg, alias, _), aggAlias, _)` when inserting the OverWindow into the call.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/call.scala:
```diff
@@ -49,6 +57,128 @@ case class Call(functionName: String, args: Seq[Expression]) extends Expression
+  * @param overWindowAlias over window alias
+  * @param overWindow      over window
+  */
+case class OverCall(
```
Hm, not sure. Would need to look into this further. But, IMO it is very important to validate that the `partitionBy` fields exist.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala:
```diff
@@ -221,6 +221,23 @@ object ProjectionTranslator {
 
+  def translateOverWindows(
+      exprs: Seq[Expression],
+      overWindows: Array[OverWindow]): Seq[Expression] = {
```
Add `tableEnv` parameter to identify whether this is a batch or a streaming query. 
In case of a streaming query, check that `overwindow.orderBy` is an `UnresolvedFieldReference` and translates `rowtime` to a `RowTime` expression and `proctime` to a `ProcTime` expression.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/OverWindowTest.scala:
```diff
@@ -0,0 +1,528 @@
+import org.apache.flink.table.utils.{StreamTableTestUtil, TableTestBase}
+import org.junit.Test
+
+class OverWindowTest extends TableTestBase {
```
I think we should add a few more tests here:
- Add two tests without following for row and range
- Check that orderBy field is correctly validated (only rowtime, proctime)
- Check that partitionBy fields are correctly validated (existing fields in the input, no complex expressions)
- Check that preceding and following are of same type
- Check that preceding and following are literals
- any other validation checks?

sunjincheng121 commented:
@fhueske Thanks a lot for your comments and suggestions. And I like the idea of move all validations into `validateInput` method. I have updated the PR with following changes:
1.Check that orderBy field is correctly validated (only rowtime, proctime)
2.Check that partitionBy fields are correctly validated (existing fields in the input, no complex expressions)
3.Check that preceding and following are of same type
4.Check that preceding and following are literals
5.Check that preceding value is > 0
6.Check that following value is >= -1
7.Move all validations into `validateInput` method.
BTW. In this PR we only add batch check in the `table.scala#window(overWindows: OverWindow*)` method.

Thanks,
SunJincheng

fhueske commented:
Hi @sunjincheng121,

thanks for the update! I was working on some refactoring of this PR as well to get the validation logic clean, but I think your approach is much nicer. 
I'll will rebase my changes on top of yours and will open a PR against your PR branch soon.

Cheers, Fabian

sunjincheng121 commented:
Hi @fhueske, Thanks for everything you did. I have merged the changes, and I have moved validations of `partitionBy` into `validateInput` method and added two tests about partitionBy:
* TestPartitionByWithUnresolved
* TestPartitionByWithNotKeyType

Thanks,
SunJincheng

fhueske commented:
Thanks for the update @sunjincheng121!

Will merge this PR.
